<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/0.0.7/segpy_slurm/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Segpy SLURM - segpy</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../css/version-select.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Segpy SLURM";
        var mkdocs_page_input_path = "segpy_slurm.md";
        var mkdocs_page_url = "/0.0.7/segpy_slurm/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> segpy
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Segpy SLURM</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#contents">Contents</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#input-data">Input data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#step-0-setup">Step 0: Setup</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#step-1-vcf-to-matrixtable">Step 1: VCF to MatrixTable</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#step-2-run-segregation">Step 2: Run segregation</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#step-3-parse-output-file">Step 3: Parse output file</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../segpy_local/">Segpy Local</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Configuration parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../output/">Output files</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../FAQ/">FAQ</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">License</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../changelog/">Changelog</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">Contributing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">Acknowledgement</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">segpy</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tutorial</li>
      <li class="breadcrumb-item active">Segpy SLURM</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="segpy-slurm">Segpy SLURM</h1>
<p>In this tutorial we illustrate how to run the Segpy pipeline on a High-Performance Computing (HPC) system using the <a href="https://slurm.schedmd.com/">SLURM</a> workload manager.</p>
<h2 id="contents">Contents</h2>
<ul>
<li><a href="#input-data">Input data</a></li>
<li><a href="#step-0-setup">Step 0: Setup</a></li>
<li><a href="#step-1-vcf-to-matrixtable">Step 1: VCF to MatrixTable</a></li>
<li><a href="#step-2-run-segregation">Step 2: Run segregation</a></li>
<li><a href="#step-3-parse-output-file">Step 3: Parse output file</a></li>
</ul>
<p>The following flowchart illustrates the steps for running the segregation analysis on an HPC system.</p>
<p align="center">
 <img src="https://github.com/user-attachments/assets/d7879700-3bba-4d53-a775-8556e3c3f6d3" width="300" height="100">
 </p>

<p><strong>Figure 1. Segpy pipeline workflow.</strong></p>
<hr />
<h3 id="input-data">Input data</h3>
<p>To execute the pipeline, users must provide two separate input files: <br></p>
<ol>
<li><strong>VCF</strong></li>
<li><strong>Pedigree file</strong></li>
</ol>
<p>The <strong>VCF</strong> should be formatted according to the standard VCF specifications, containing information about genetic variants, including their positions, alleles, and genotype information for each individual in the study. <br></p>
<p>The <strong>Pedigree file</strong> should be in .ped format, structured such that each line describes an individual in the study and includes the following columns: <code>familyid</code>, <code>individualid</code>, <code>parentalid</code>, <code>maternalid</code>, <code>sex</code>, <code>phenotype</code>.</p>
<ul>
<li>The <code>familyid</code> column must contain identifiers for the family.</li>
<li>The <code>individualid</code> column must contain identifiers for the individual that match the VCF file.</li>
<li>The <code>parentalid</code> column must contain identifiers for the father (0 if unknown).</li>
<li>The <code>maternalid</code> column must contain identifiers for the mother (0 if unknown).</li>
<li>The <code>sex</code> column must describe the biological sex of the individual (1 = male, 2 = female, 0 = unknown).</li>
<li>The <code>phenotype</code> column must describe the phenotypic data (1 = unaffected, 2 = affected, -9 = missing).</li>
</ul>
<p>We provide an example .ped file <a href="https://github.com/neurobioinfo/segpy/blob/main/test/data/iPSC_2.ped">HERE</a>. </p>
<p><strong>NOTE:</strong> For Case-control analyses, the <code>familyid</code> column should be 'case' for affected individuals and 'control' for unaffected individuals. Furthermore, the <code>parentalid</code> and <code>maternalid</code> can be '0' for every listed individual in case-control analyses. </p>
<hr />
<h3 id="step-0-setup">Step 0: Setup</h3>
<p>Prior to initiating the pipeline, users must provide the paths to:</p>
<ol>
<li>The directory containing <code>segpy.pip</code> (PIPELINE_HOME)</li>
<li>The directory designated for the pipeline's outputs (PWD)</li>
<li>The VCF file (VCF)</li>
<li>The pedigree file (PED)</li>
</ol>
<p>Use the following code to define these paths:</p>
<pre><code># Define necessary paths
export PIPELINE_HOME=path/to/segpy.pip
PWD=path/to/outfolder
VCF=path/to/VCF.vcf
PED=path/to/pedigree.ped
</code></pre>
<p>To ensure that <code>segpy.pip</code> was defined properly, run the following command:</p>
<pre><code>module load apptainer/1.2.4

bash $PIPELINE_HOME/launch_segpy.sh -h
</code></pre>
<p>Which should return the following:</p>
<pre><code>------------------------------------ 
segregation pipeline version 0.0.6 is loaded

------------------- 
Usage:  segpy.pip/launch_segpy.sh [arguments]
        mandatory arguments:
                -d  (--dir)      = Working directory (where all the outputs will be printed) (give full path) 
                -s  (--steps)      = Specify what steps, e.g., 2 to run just step 2, 1-3 (run steps 1 through 3). 'ALL' to run all steps.
                                steps:
                                0: initial setup
                                1: create hail matrix
                                2: run segregation
                                3: final cleanup and formatting

        optional arguments:
                -h  (--help)      = Get the program options and exit.
                --jobmode  = The default for the pipeline is local. If you want to run the pipeline on slurm system, use slurm as the argument. 
                --analysis_mode  = The default for the pipeline is analysing single or multiple family. If you want to run the pipeline on case-control, use case-control as  the argumnet. 
                --parser             = 'general': to general parsing, 'unique': drop multiplicities 
                -v  (--vcf)      = VCF file (mandatory for steps 1-3)
                -p  (--ped)      = PED file (mandatory for steps 1-3)
                -c  (--config)      = config file [CURRENT: &quot;/scratch/fiorini9/segpy.pip/configs/segpy.config.ini&quot;]
                -V  (--verbose)      = verbose output 

 ------------------- 
 For a comprehensive help, visit  https://neurobioinfo.github.io/segpy/latest/ for documentation. 
</code></pre>
<p>Once the necessary paths have been defined, we can initialize the pipeline using the following command:</p>
<pre><code>module load apptainer/1.2.4

bash $PIPELINE_HOME/launch_segpy.sh \
-d $PWD \
--steps 0 \
--analysis_mode single_family \
--jobmode slurm

</code></pre>
<p>Where <code>analysis_mode</code> is one of <code>single_family</code>, <code>multiple_family</code>, or <code>case-control</code> depending on the study design. </p>
<p>After running this command, the working directory (PWD) should have the following structure:</p>
<pre><code>PWD
├── configs
│   └── segpy.config.ini
├── launch_summary_log.txt
└── logs
    ├── jobs
    └── spark
</code></pre>
<p>The <code>configs</code> directory contains a .ini file with adjustable parameters for the pipeline. Please see the <a href="../reference/">Configuration parameters</a> section of this documentation for more information regarding the adjustable parameters. The <code>logs</code> directory documents the parameters and outputs from each analytical step of the Segpy pipeline to ensure reproducibility. The <code>launch_summary_log.txt</code> is a cumulative documentation of all submitted jobs throughout the analysis.</p>
<p>After completing Step 0, open the <code>segpy.config.ini</code> file to adjust the general parameters for the analysis. The following parameters must be adjusted:</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th style="text-align: left;">Default</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr>
<td>CONTAINER_MODULE</td>
<td style="text-align: left;">CONTAINER_MODULE='apptainer/1.2.4'</td>
<td style="text-align: left;">The version of Singularity/Apptainer loaded onto your system.</td>
</tr>
<tr>
<td>ACCOUNT</td>
<td style="text-align: left;">ACCOUNT=user</td>
<td style="text-align: left;">Your SLURM user account.</td>
</tr>
<tr>
<td>CSQ</td>
<td style="text-align: left;">CSQ=TRUE</td>
<td style="text-align: left;">Whether or not to include variant annotations from VEP in the output file.</td>
</tr>
<tr>
<td>GRCH</td>
<td style="text-align: left;">GRCH=GRCh38</td>
<td style="text-align: left;">Reference genome version (GRCh38 or GRCh37).</td>
</tr>
<tr>
<td>AFFECTEDS_ONLY</td>
<td style="text-align: left;">AFFECTEDS_ONLY=FALSE</td>
<td style="text-align: left;">Whether or not to only include families with at least one affected individual.</td>
</tr>
<tr>
<td>FILTER_VARIANT</td>
<td style="text-align: left;">FILTER_VARIANT=TRUE</td>
<td style="text-align: left;">Filter the output file using relevant counting column values where 'fam_aff_vrt'+'fam_aff_homv'+'fam_naf_vrt'+'fam_naf_homv' &gt;0. See <a href="#step-3-parse-output-file">Step 3: Parse output file</a> for more information.</td>
</tr>
<tr>
<td>JAVATOOLOPTIONS</td>
<td style="text-align: left;">JAVATOOLOPTIONS="-Xmx6g"</td>
<td style="text-align: left;">Java Virtual Machine (JVM) configuration setting. For most use cases the default value will be appropriate.</td>
</tr>
</tbody>
</table>
<p>Use the following code to modify the <code>segpy.config.ini</code> file:</p>
<pre><code># 1) Open .ini file to edit with nano
nano $PWD/configs/segpy.config.ini

# 2) Make appropriate changes.

# 3) Save and exit the .ini file with ctrl+0, enter ,ctrl+x
</code></pre>
<p><strong>NOTE</strong>: If you are using Segpy SLURM you must uncomment the <code>CONTAINER_MODULE</code> parameter. </p>
<hr />
<h3 id="step-1-vcf-to-matrixtable">Step 1: VCF to MatrixTable</h3>
<p>In step 1, we will convert the user-provided VCF file to the <a href="https://hail.is/docs/0.2/overview/matrix_table.html">Hail MatrixTable</a> format, which is designed to efficiently store and manipulate large-scale genomic datasets.</p>
<p>Prior to running step 1, open the <code>segpy.config.ini</code> file to adjust the job submission parameters for step 1:</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th style="text-align: left;">Default</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr>
<td>WALLTIME_ARRAY["step1"]</td>
<td style="text-align: left;">WALLTIME_ARRAY["step1"]=00-5:00</td>
<td style="text-align: left;">Number of CPUs for the step 1 job sumission.</td>
</tr>
<tr>
<td>THREADS_ARRAY["step1"]</td>
<td style="text-align: left;">THREADS_ARRAY["step1"]=8</td>
<td style="text-align: left;">Amount of memory (RAM) for the step 1 job sumission.</td>
</tr>
<tr>
<td>MEM_ARRAY["step1"]</td>
<td style="text-align: left;">MEM_ARRAY["step1"]=10g</td>
<td style="text-align: left;">Amount of time for the step 1 job sumission.</td>
</tr>
</tbody>
</table>
<p>Use the following code to modify the <code>segpy.config.ini</code> file:</p>
<pre><code># 1) Open .ini file to edit with nano
nano $PWD/configs/segpy.config.ini

# 2) Make appropriate changes.

# 3) Save and exit the .ini file with ctrl+0, enter ,ctrl+x
</code></pre>
<p>Once the parameters have been adjusted, run step 1 using the following command:</p>
<pre><code>bash $PIPELINE_HOME/launch_segpy.sh \
-d $PWD \
--steps 1 \
--vcf $VCF
</code></pre>
<p>After running this command, a <code>step1</code> directory will be created in the working directory (PWD), which will contain the output files from Hail. The <code>step1</code> directory should have the following structure:</p>
<pre><code>step1
└── VEP_iPSC.mt
    ├── cols
    │   ├── metadata.json.gz
    │   ├── README.txt
    │   ├── rows
    │   │   ├── metadata.json.gz
    │   │   └── parts
    │   │       └── part-0
    │   └── _SUCCESS
    ├── entries
    │   ├── metadata.json.gz
    │   ├── README.txt
    │   ├── rows
    │   │   ├── metadata.json.gz
    │   │   └── parts
    │   │       ├── part-0-5eb5d5aa-c7b2-4acf-bea1-b55e6f4d9bff
    │   │       ├── part-1-8ba09e12-618c-416b-8abe-8dd858760e2f
    │   │       └── part-2-7b8e9e61-db1b-4697-8e27-c797f9082545
    │   └── _SUCCESS
    ├── globals
    │   ├── globals
    │   │   ├── metadata.json.gz
    │   │   └── parts
    │   │       └── part-0
    │   ├── metadata.json.gz
    │   ├── README.txt
    │   ├── rows
    │   │   ├── metadata.json.gz
    │   │   └── parts
    │   │       └── part-0
    │   └── _SUCCESS
    ├── index
    │   ├── part-0-5eb5d5aa-c7b2-4acf-bea1-b55e6f4d9bff.idx
    │   │   ├── index
    │   │   └── metadata.json.gz
    │   ├── part-1-8ba09e12-618c-416b-8abe-8dd858760e2f.idx
    │   │   ├── index
    │   │   └── metadata.json.gz
    │   └── part-2-7b8e9e61-db1b-4697-8e27-c797f9082545.idx
    │       ├── index
    │       └── metadata.json.gz
    ├── metadata.json.gz
    ├── README.txt
    ├── references
    ├── rows
    │   ├── metadata.json.gz
    │   ├── README.txt
    │   ├── rows
    │   │   ├── metadata.json.gz
    │   │   └── parts
    │   │       ├── part-0-5eb5d5aa-c7b2-4acf-bea1-b55e6f4d9bff
    │   │       ├── part-1-8ba09e12-618c-416b-8abe-8dd858760e2f
    │   │       └── part-2-7b8e9e61-db1b-4697-8e27-c797f9082545
    │   └── _SUCCESS
    └── _SUCCESS
</code></pre>
<p>A deatiled description of the outputs can be found in the <a href="https://hail.is/docs/0.2/index.html">Hail documentation</a>.</p>
<hr />
<h3 id="step-2-run-segregation">Step 2: Run segregation</h3>
<p>In step 2, we will leverage the Hail ouputs from step 1 to perform segregation analysis based on the information described in the user-provided pedigree file.  </p>
<p>Prior to running step 2, open the segpy.config.ini file to adjust the job submission parameters for step 2:</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th style="text-align: left;">Default</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr>
<td>WALLTIME_ARRAY["step2"]</td>
<td style="text-align: left;">WALLTIME_ARRAY["step2"]=00-5:00</td>
<td style="text-align: left;">Number of CPUs for the step 2 job sumission.</td>
</tr>
<tr>
<td>THREADS_ARRAY["step2"]</td>
<td style="text-align: left;">THREADS_ARRAY["step2"]=8</td>
<td style="text-align: left;">Amount of memory (RAM) for the step 2 job sumission.</td>
</tr>
<tr>
<td>MEM_ARRAY["step2"]</td>
<td style="text-align: left;">MEM_ARRAY["step2"]=10g</td>
<td style="text-align: left;">Amount of time for the step 2 job sumission.</td>
</tr>
</tbody>
</table>
<p>Use the following code to modify the <code>segpy.config.ini</code> file:</p>
<pre><code># 1) Open .ini file to edit with nano
nano $PWD/configs/segpy.config.ini

# 2) Make appropriate changes.

# 3) Save and exit the .ini file with ctrl+0, enter ,ctrl+x
</code></pre>
<p>Once the parameters have been adjusted, run step 2 using the following command:</p>
<pre><code>bash $PIPELINE_HOME/launch_segpy.sh \
-d $PWD \
--steps 2 \
--vcf $VCF
--ped $PED
</code></pre>
<p>After running this command, a <code>step2</code> directory will be created in the working directory (PWD), which will contain the output files from the segregation analysis. The <code>step2</code> directory should have the following structure:</p>
<pre><code>step2
├── finalseg.csv
└── temp
</code></pre>
<p>The <code>finalseg.csv</code> file contains the variant counts obtained from the segregation analysis, as well as the variant annotations supplied by VEP. Please see the <a href="../output/">Output files</a> section of this documentation for a comprehensive description of the file contents.  </p>
<hr />
<h3 id="step-3-parse-output-file">Step 3: Parse output file</h3>
<p>In step 3, we will parse the <code>finalseg.csv</code> file obtained from step 2 to simplify it and reduce its size.</p>
<p>Prior to running step 3, open the segpy.config.ini file to adjust the job submission parameters for step 3:</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th style="text-align: left;">Default</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr>
<td>WALLTIME_ARRAY["step3"]</td>
<td style="text-align: left;">WALLTIME_ARRAY["step3"]=00-5:00</td>
<td style="text-align: left;">Number of CPUs for the step 3 job sumission.</td>
</tr>
<tr>
<td>THREADS_ARRAY["step3"]</td>
<td style="text-align: left;">THREADS_ARRAY["step3"]=8</td>
<td style="text-align: left;">Amount of memory (RAM) for the step 3 job sumission.</td>
</tr>
<tr>
<td>MEM_ARRAY["step3"]</td>
<td style="text-align: left;">MEM_ARRAY["step3"]=10g</td>
<td style="text-align: left;">Amount of time for the step 3 job sumission.</td>
</tr>
</tbody>
</table>
<p>Use the following code to modify the <code>segpy.config.ini</code> file:</p>
<pre><code># 1) Open .ini file to edit with nano
nano $PWD/configs/segpy.config.ini

# 2) Make appropriate changes.

# 3) Save and exit the .ini file with ctrl+0, enter ,ctrl+x
</code></pre>
<p>Once the parameters have been adjusted, we can run step 3. If you simply want to remove unnecessary characters  (e.g., <code>"</code> and <code>[ ]</code>) use the following command:</p>
<pre><code>bash $PIPELINE_HOME/launch_segpy.sh \
-d $PWD \
--steps 3 \
--parser general
</code></pre>
<p>If you want to eliminate duplicated variant entries resulting from VEP annotations, use the following command:</p>
<pre><code>bash $PIPELINE_HOME/launch_segpy.sh \
-d $PWD \
--steps 3 \
--parser unique
</code></pre>
<p>After running this command, a <code>step3</code> directory will be created in the working directory (PWD), which will contain the output files from the segregation analysis. The <code>step3</code> directory should have the following structure:</p>
<pre><code>step3
└── finalseg_cleaned_general.csv
└── finalseg_cleaned_unique.csv
</code></pre>
<p>The <code>finalseg_cleaned_general.csv</code> and <code>finalseg_cleaned_unique.csv</code> files contain the parsed variant counts obtained from the segregation analysis, as well as the variant annotations supplied by VEP. Please see the <a href="../output/">Output files</a> section of this documentation for a comprehensive description of the file contents.  </p>
<hr />
<p><strong>Note</strong>: You can execute steps 1 to 3 sequentially, using the following command:</p>
<pre><code>bash $PIPELINE_HOME/launch_segpy.sh \
-d $PWD \
--steps 1-3 \
--vcf $VCF \
--ped $PED \
--parser general
</code></pre>
<p><strong><a href="#segpy-slurm">⬆ back to top</a></strong></p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../installation/" class="btn btn-neutral float-left" title="Installation"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../segpy_local/" class="btn btn-neutral float-right" title="Segpy Local">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../installation/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../segpy_local/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
      <script src="../js/version-select.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
